### PR TITLE
Modded River biomes

### PR DESCRIPTION
**Pull request by valoeghese**
Simple change that lets modders specify river biomes through overriding a method in your Biome's class
Being able to specify river biomes is very useful for biome mods. It means the said mods can:

- Set the rivers in their snowy biomes to be frozen
- Remove rivers in biomes in which they look unusual
- Add custom rivers to themed biomes

The code has been tested and it works.

Please note any changes I should make to this pr. Thanks.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/BiomeManager.java [hunk 1]
```diff
@@ -21,222 +21,243 @@
...
 import net.minecraft.util.WeightedRandom;
 import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.Biomes;
 import net.minecraft.world.biome.provider.BiomeProvider;
 
-import javax.annotation.Nullable;
-
 public class BiomeManager
 {
-    private static TrackedList<BiomeEntry>[] biomes = setupBiomes();
```
Comment (valoeghese) [inline src/main/java/net/minecraftforge/common/BiomeManager.java]: This unusual change is my IDE randomly deciding to reindent stuff

**Comment by LexManos** [comment]
> It seems my IDE has changed the indent size of some files so the change looks bigger than it really is 

Do not submit formatting changes...

**Comment by valoeghese** [comment]
Eclipse is weird. I'll need to change the formatting settings for my ide, fix it, then update the pr.

**Comment by valoeghese** [comment]
Where can I find the specifications for formatting in forge?

**Comment by valoeghese** [comment]
Reverting formatting (to what you had previously)

**Comment by valoeghese** [comment]
This small commit should have reverted whitespacing and import order to the forge standard

File: src/main/java/net/minecraftforge/common/MinecraftForge.java [hunk 2]
```diff
@@ -19,9 +19,9 @@
 
 package net.minecraftforge.common;
 
+import net.minecraft.crash.CrashReport;
 import net.minecraftforge.eventbus.api.BusBuilder;
 import net.minecraftforge.eventbus.api.IEventBus;
-import net.minecraft.crash.CrashReport;
```
Comment (valoeghese) [inline src/main/java/net/minecraftforge/common/MinecraftForge.java]: reverting these remaining formatting things now

**Comment by valoeghese** [comment]
> It seems my IDE has changed the indent size of some files so the change looks bigger than it really is
> 
> Do not submit formatting changes...

Fixed

File: patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
++++ b/net/minecraft/world/gen/layer/MixRiverLayer.java
+@@ -22,7 +22,10 @@
+       if (LayerUtil.func_202827_a(i)) {
+          return i;
+       } else if (j == field_202725_h) {
+-         if (i == field_202721_d) {
++    	 net.minecraft.world.biome.Biome modded_river_biome = net.minecraftforge.common.BiomeManager.getRiver(Registry.field_212624_m.func_148745_a(i));
++         if (modded_river_biome != null) {
++        	return Registry.field_212624_m.func_148757_b(modded_river_biome);
++         } else if (i == field_202721_d) {
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch]: 1. That's not how you name variables in Java.
2. Don't patch the if into an elseif. Your if already returns, so you don't need an else-if. That makes this a 2-line patch (`if (moddedRiverBiome != null) return ...`

File: src/main/java/net/minecraftforge/common/BiomeManager.java [hunk 4]
```diff
@@ -123,6 +138,12 @@ public static void removeBiome(BiomeType type, BiomeEntry entry)
         return list != null ? ImmutableList.copyOf(list) : null;
     }
 
+    @Nullable
+    public static Biome getRiver(Biome biome)
+    {
+        return modded_river_map.getOrDefault(biome, null);
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/BiomeManager.java]: Why not just `get`?

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch (hunk 3 above)
Comment (valoeghese) [reply patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch]: Thanks for pointing that out! Will change that

Comment (valoeghese) [reply patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch]: done

File: src/main/java/net/minecraftforge/common/BiomeManager.java (hunk 4 above)
Comment (valoeghese) [reply src/main/java/net/minecraftforge/common/BiomeManager.java]: done :)

**Comment by valoeghese** [comment]
Updated PR to include requested changes by diesieben07.

File: patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch [hunk 5]
```diff
@@ -0,0 +1,11 @@
+--- a/net/minecraft/world/gen/layer/MixRiverLayer.java
++++ b/net/minecraft/world/gen/layer/MixRiverLayer.java
+@@ -22,6 +22,8 @@
+       if (LayerUtil.func_202827_a(i)) {
+          return i;
+       } else if (j == field_202725_h) {
++    	 net.minecraft.world.biome.Biome moddedRiverBiome = net.minecraftforge.common.BiomeManager.getRiver(Registry.field_212624_m.func_148745_a(i));
++         if (moddedRiverBiome != null) return Registry.field_212624_m.func_148757_b(moddedRiverBiome);
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch]: Indentation is wrong here. There's a tab in there somewhere.

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch (hunk 5 above)
Comment (valoeghese) [reply patches/minecraft/net/minecraft/world/gen/layer/MixRiverLayer.java.patch]: fixed

**Comment by valoeghese** [comment]
fixed remaining indentation problems

**Comment by LexManos** [comment]
This being a hash map is rather expensive just for world gen.
Why not make it a getBeach() function on the biome?
Either that, or keep it as a id->id map so that we only have one map lookup.
Tho that would require rebuilding the map on registry rebaking.

**Comment by valoeghese** [comment]
You mean that I should move rivers to a `Biome#getRiver()` method?

**Comment by Wyn-Price** [comment]
Wouldn't a simple `public Biome riverBiome` field on the Biome class be sufficient?

**Comment by LexManos** [comment]
Thats what I just said... But instead of a public direct field where anyone can edit. Add a getter.

**Comment by valoeghese** [comment]
Moved modded rivers to the biome class as requested (`Biome#getRiver()`)

**Comment by Wyn-Price** [comment]
My idea for a field was for if you wanted to change a vanilla biomes river biomes but doing a registry replacement is fine

**Comment by valoeghese** [comment]
If anyone could provide further feedback on this PR it would be appreciated :)

**Pull request merged**